synapse/handlers/ui_auth/checkers.py

```diff
@@ -237,11 +237,65 @@ async def check_auth(self, authdict: dict, clientip: str) -> Any:
         return await self._check_threepid("msisdn", authdict)
 
 
+class RegistrationTokenAuthChecker(UserInteractiveAuthChecker):
+    AUTH_TYPE = LoginType.REGISTRATION_TOKEN
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__(hs)
+        self._enabled = bool(hs.config.registrations_require_token)
+        self.store = hs.get_datastore()
+
+    def is_enabled(self) -> bool:
+        return self._enabled
+
+    async def check_auth(self, authdict: dict, clientip: str) -> Any:
+        if "token" not in authdict:
+            raise LoginError(400, "Missing token", Codes.MISSING_PARAM)
+        if not isinstance(authdict["token"], str):
+            raise LoginError(400, "Token must be a string", Codes.INVALID_PARAM)
+        if "session" not in authdict:
+            raise SynapseError(400, "Missing UIA session", Codes.MISSING_PARAM)
+        session = authdict["session"]
+
+        # If the LoginType.REGISTRATION_TOKEN stage has already been completed,
+        # return early to avoid incrementing `pending` again.
+        # TODO: check stored token matches provided token?
```

Comment:
I think it'd be confusing if a client presented two different tokens, both with a successful result, and upon completing registration finding that the first token was taken.

Erroring on switching tokens (or overwriting the old token, though this requires some extra book-keeping) is probably clearer from the client's perspective.

Reply:
Yep, I've now got it raising a `400` if the token changes. I think it's probably not worth overwriting old tokens (involving changing pending) given that clients are (hopefully) unlikely to retry a stage they have already successfully completed.